keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1603 @@
+`nft` command lines without the leading "`nft`" (which are then
+executed atomically):
+
+```
+add table ip kube_proxy { comment "Kubernetes service proxying rules"; }
+add chain ip kube_proxy services
+add rule ip kube_proxy services ip daddr . ip protocol . th dport vmap @service_ips
+```
+
+The "object" form is more logical and easy to understand, but the
+"command" form is better for dynamic usage. In particular, it allows
+you to add and remove individual chains, rules, map/set elements, etc,
+without needing to also include the chains/rules/elements that you are
+not modifying.
+
+The examples below all show the "object" form of data, but it should
+be understood that these are examples of what would be seen in `nft
+list` output after kube-proxy creates the rules (with additional
+`#`-preceded comments added to help the KEP reader), not examples of
+the data we will actually be passing to `nft`.
+
+The examples below are also all IPv4-specific, for simplicity. When
+actually writing out rules for nft, we will need to switch between,
+e.g., "`ip daddr`" and "`ip6 daddr`" appropriately, to match an IPv4
+or IPv6 destination address. This will actually be fairly simple
+because the `nft` command lets you create "variables" (really
+constants) and substitute their values into the rules. Thus, we can
+just always have the rule-generating code write "`$IP daddr`", and
+then pass either "`-D IP=ip`" or "`-D IP=ip6`" to `nft` to fix it up.)
+
+(Also, most of the examples below have not actually been tested and
+may have syntax errors. Caveat lector.)
+
+#### Versioning and compatibility
+
+Since nftables is subject to much more development than iptables has
+been recently, we will need to pay more attention to kernel and tool
+versions.
+
+The `nft` command has a `--check` option which can be used to check if
+a command could be run successfully; it parses the input, and then
+(assuming success), uploads the data to the kernel and asks the kernel
+to check it (but not actually act on it) as well. Thus, with a few
+`nft --check` runs at startup we should be able to confirm what
+features are known to both the tooling and the kernel.
+
+It is not yet clear what the minimum kernel or `nft` command-line
+versions needed by the `nftables` backend will be. The newest feature
+used in the examples below was added in Linux 5.6, released in March
+2020 (though they could be rewritten to not need that feature).
+
+It is possible some users will not be able to upgrade from the
+`iptables` and `ipvs` backends to `nftables`. (Certainly the
+`nftables` backend will not support RHEL 7, which some people are
+still using Kubernetes with.)
+
+#### NAT rules
+
+##### General Service dispatch
+
+For ClusterIP and external IP services, we will use an nftables
+"verdict map" to store the logic about where to dispatch traffic,
+based on destination IP, protocol, and port. We will then need only a
+single actual rule to apply the verdict map to all inbound traffic.
```

Comment:
let me see if I understand, right now the Service. logic is O(N) , so you iterate every single rules, does this means that this verdict map will make this O(1)?

Reply:
Yes, O(1) on the "data plane" side. On the "control plane" side, adding/removing elements from a vmap isn't O(1) because there's sanity-checking (eg making sure that you haven't created a loop with `jump`/`goto` rules). It's not clear how big that cost is at large-k8s-cluster scale (though _at worst_ it would still be doing less work than iptables with `MinimizeIPTablesRestore`)

Reply:
will be interesting to see the output of the tests 👍